File: zmq/Go/zmq_simple_publisher.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"strings"
	"time"

	zmq "github.com/pebbe/zmq4"
)

func main() {
	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("GO ZMQ SIMPLE PUBLISHER")
	fmt.Println(strings.Repeat("=", 60))

	// Step 1: Create PUB socket
	// Go handles context implicitly in NewSocket usually, or you can create one.
	// For simplicity, we use the default context.
	fmt.Println("\n[Step 1] Creating PUB socket...")
	publisher, err := zmq.NewSocket(zmq.PUB)
	if err != nil {
		log.Fatal(err)
	}
	// defer ensures the socket closes when main function exits
	defer publisher.Close()
	fmt.Println("✓ PUB socket created")

	// Step 2: Bind
	address := "tcp://*:5555"
	fmt.Printf("\n[Step 2] Binding to %s...\n", address)
	err = publisher.Bind(address)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("✓ Publisher bound to %s\n", address)

	// Step 3: Wait
	fmt.Println("\n[Step 3] Waiting 2 seconds for subscribers to connect...")
	time.Sleep(2 * time.Second)
	fmt.Println("✓ Ready to publish!")

	// Step 4: Loop
	fmt.Println("\n[Step 4] Publishing messages (Press Ctrl+C to stop)...")
	fmt.Println(strings.Repeat("-", 60))

	msgCount := 0
	for {
		msgCount++
		// Generate random temperature
		temperature := rand.Intn(16) + 15 // 15-30
		msg := fmt.Sprintf("Temperature: %d°C", temperature)

		// Send message (0 means no special flags)
		_, err := publisher.Send(msg, 0)
		if err != nil {
			log.Println("Error sending:", err)
			break
		}

		// Log
		timestamp := time.Now().Format("15:04:05")
		fmt.Printf("[%s] Sent message #%d: %s\n", timestamp, msgCount, msg)

		time.Sleep(1 * time.Second)
	}
}

File: zmq/python/zmq_simple_publisher.py
#!/usr/bin/env python3
"""
Simple ZMQ Publisher Example
This script publishes messages to subscribers
"""
import zmq
import time
import random

def main():
    print("=" * 60)
    print("ZMQ SIMPLE PUBLISHER")
    print("=" * 60)
    
    # Step 1: Create ZMQ context
    print("\n[Step 1] Creating ZMQ context...")
    context = zmq.Context()
    print("✓ Context created")
    
    # Step 2: Create PUB socket
    print("\n[Step 2] Creating PUB socket...")
    socket = context.socket(zmq.PUB)
    print("✓ PUB socket created")
    
    # Step 3: Bind to address
    address = "tcp://*:5555"
    print(f"\n[Step 3] Binding to {address}...")
    socket.bind(address)
    print(f"✓ Publisher bound to {address}")
    print("  (Listening on port 5555, accepting connections from subscribers)")
    
    # Step 4: Wait a moment for subscribers to connect
    print("\n[Step 4] Waiting 2 seconds for subscribers to connect...")
    time.sleep(2)
    print("✓ Ready to publish!")
    
    # Step 5: Publish messages
    print("\n[Step 5] Publishing messages (press Ctrl+C to stop)...")
    print("-" * 60)
    
    try:
        message_count = 0
        while True:
            # Create a simple message
            temperature = random.randint(15, 30)
            message = f"Temperature: {temperature}°C"
            
            # Send message
            socket.send_string(message)
            message_count += 1
            
            # Log
            timestamp = time.strftime("%H:%M:%S")
            print(f"[{timestamp}] Sent message #{message_count}: {message}")
            
            # Wait before sending next message
            time.sleep(1)
            
    except KeyboardInterrupt:
        print("\n" + "-" * 60)
        print(f"\n[Shutdown] Received interrupt signal")
        print(f"  Total messages sent: {message_count}")
    
    finally:
        # Step 6: Cleanup
        print("\n[Step 6] Cleaning up...")
        socket.close()
        print("✓ Socket closed")
        context.term()
        print("✓ Context terminated")
        print("\n" + "=" * 60)
        print("Publisher stopped successfully")
        print("=" * 60)

if __name__ == "__main__":
    main()

